pkg/sync/syncWorker.go

```diff
@@ -0,0 +1,205 @@
+		nodeRegistry: nodeRegistry,
+		registrant:   registrant,
+		store:        store,
+		wg:           sync.WaitGroup{},
+	}
+	if err := s.start(); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (s *syncWorker) start() error {
+	nodes, err := s.nodeRegistry.GetNodes()
+	if err != nil {
+		return err
+	}
+	for _, node := range nodes {
+		if node.NodeID == s.registrant.NodeID() || !node.IsHealthy {
+			continue
+		}
+		s.subscribeToNode(node)
+	}
+	return nil
+}
+
+func (s *syncWorker) close() {
+	s.wg.Wait()
+}
+
+func (s *syncWorker) subscribeToNode(node registry.Node) {
+	tracing.GoPanicWrap(
+		s.ctx,
+		&s.wg,
+		fmt.Sprintf("node-subscribe-%d", node.NodeID),
+		func(ctx context.Context) {
+			var err error
+			for {
+				if err != nil {
+					log.Error(fmt.Sprintf("Error: %v, retrying...", err))
+					time.Sleep(1 * time.Second)
+				}
+				err = nil
+
+				conn, err := s.connectToNode(node)
+				if err != nil {
+					continue
+				}
+				stream, err := s.setupStream(node, conn)
+				if err != nil {
+					continue
+				}
+				err = s.listenToStream(stream)
+				if err != nil {
+					continue
+				}
+			}
+		},
+	)
+}
+
+func (s *syncWorker) connectToNode(node registry.Node) (*grpc.ClientConn, error) {
+	addr := node.HttpAddress
+	log.Info(fmt.Sprintf("attempting to connect to %s", addr))
+	conn, err := grpc.NewClient(
+		addr,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultCallOptions(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to connect to peer: %v", err)
+	}
+	log.Info(fmt.Sprintf("Successfully connected to peer at %s", addr))
+	return conn, nil
+}
+
+func (s *syncWorker) setupStream(node registry.Node, conn *grpc.ClientConn) (message_api.ReplicationApi_BatchSubscribeEnvelopesClient, error) {
+	client := message_api.NewReplicationApiClient(conn)
+	stream, err := client.BatchSubscribeEnvelopes(
+		s.ctx,
+		&message_api.BatchSubscribeEnvelopesRequest{
+			Requests: []*message_api.BatchSubscribeEnvelopesRequest_SubscribeEnvelopesRequest{
+				{
+					Query: &message_api.EnvelopesQuery{
+						Filter: &message_api.EnvelopesQuery_OriginatorNodeId{
+							OriginatorNodeId: node.NodeID,
+						},
+						LastSeen: nil,
+					},
+				},
+			},
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"Failed to batch subscribe to peer: %v",
+			err,
+		)
+	}
+	return stream, nil
+}
+
+func (s *syncWorker) listenToStream(stream message_api.ReplicationApi_BatchSubscribeEnvelopesClient) error {
+	for {
+		envs, err := stream.Recv()
+		// TODO(rich) Handle normal stream closure properly
+		if err != nil {
+			return fmt.Errorf(
+				"Stream closed with error: %v",
+				err)
+		}
+		for _, env := range envs.Envelopes {
+			s.insertEnvelope(env)
+		}
+	}
+}
+
+func (s *syncWorker) insertEnvelope(env *message_api.OriginatorEnvelope) {
+	log.Info(fmt.Sprintf("Replication server received envelope %s", env))
+	// TODO(rich) Validation logic - share code with API service and publish worker
```

Comment:
I've been meaning to make some nice utility functions and wrappers for these anyways. I can take that TODO